### PR TITLE
[TOK-337] Allow builder receiver address update

### PR DESCRIPTION
**Pull request by jmendiola222**
## What

Allow builder to update his receiver address, by submiting a new address, and witing for the KYCApprover to accept it.

## Why

It might be for organization purposes, lost of PK or even compromise. Either way, we should allow Builder to change it.

## Testing

By calling the submit request and kycApproval afterwars

## Refs

- https://rsklabs.atlassian.net/browse/TOK-337

### CODE REVIEW

File: src/BuilderRegistryRootstockCollective.sol [hunk 1]
```diff
@@ -152,6 +158,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
     // ---- External Functions -----
     // -----------------------------
 
+    /**
+     * @notice Builder submites a request to replace his rewardReceiver address,
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:162]:
```suggestion
     * @notice Builder submits a request to replace his rewardReceiver address,
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 2]
```diff
@@ -491,6 +550,20 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
         emit GaugeCreated(builder_, address(gauge_), msg.sender);
     }
 
+    /**
+     * @notice Builder submites a request to replace his rewardReceiver address,
+     * the request will then need to be appoved by `approveBuidlerRewardReceiverReplacement`
+     * @dev reverts if Buider is not Operational
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:556]:
```suggestion
     * @dev reverts if Builder is not Operational
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 3]
```diff
@@ -491,6 +550,20 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+     * @param newRewardReceiver_ new address the builder is requesting to use
+     */
+    function _submitRewardReceiverReplacementRequest(address newRewardReceiver_) internal {
+        // Only builder can submit a reward receiver address replacement
+        address _builder = msg.sender;
+        // Only oprational builder can initiate this action
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:562]:
```suggestion
        // Only operational builder can initiate this action
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 4]
```diff
@@ -491,6 +550,20 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+        if (!isBuilderOperational(_builder)) revert NotOperational();
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:563]: What's the reasoning behind not allowing a paused builder to update the receiver address? Let me double-check this with product.

File: src/BuilderRegistryRootstockCollective.sol [hunk 5]
```diff
@@ -152,6 +158,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
     // ---- External Functions -----
     // -----------------------------
 
+    /**
+     * @notice Builder submites a request to replace his rewardReceiver address,
+     * the request will then need to be appoved by `approveBuidlerRewardReceiverReplacement`
+     * @dev reverts if Buider is not Operational
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:164]:
```suggestion
     * @dev reverts if Builder is not Operational
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 6]
```diff
@@ -152,6 +158,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+     * @param newRewardReceiver_ new address the builder is requesting to use
+     */
+    function submitRewardReceiverReplacementRequest(address newRewardReceiver_) external {
+        _submitRewardReceiverReplacementRequest(newRewardReceiver_);
+        emit BuilderRewardReceiverReplacementRequested(msg.sender, newRewardReceiver_);
+    }
+
+    /**
+     * @notice Builder cancels his request to replace his rewardReceiver address
+     * @dev reverts if Buider is not Operational
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:174]:
```suggestion
     * @dev reverts if Builder is not Operational
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 7]
```diff
@@ -152,6 +158,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+     */
+    function cancelRewardReceiverReplacementRequest() external {
+        // By overriding the replacement with the current, it efectily cancels the request
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:177]:
```suggestion
        // By overriding the replacement with the current, it effectively cancels the request
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 8]
```diff
@@ -152,6 +158,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+        address _currRewardReceiver = builderRewardReceiver[msg.sender];
+        _submitRewardReceiverReplacementRequest(_currRewardReceiver);
+        emit BuilderRewardReceiverReplacementCancelled(msg.sender, _currRewardReceiver);
+    }
+
+    /**
+     * @notice KYCApprover approves Builder's request to replace his rewardReceiver address
+     * @dev reverts is provided `rewardReceiverReplacement_` don't not matched Builder's request
```
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol:185]:
```suggestion
     * @dev reverts if provided `rewardReceiverReplacement_` doesn't match Builder's request
```

File: src/interfaces/IBackersManagerRootstockCollective.sol [hunk 9]
```diff
@@ -20,6 +20,11 @@ interface IBackersManagerRootstockCollective {
      */
     function builderRewardReceiver(address builder_) external view returns (address rewardReceiver_);
 
+    /**
+     * @notice returns true, if the builder has an opened request to replace his receiver address
+     */
+    function hasBuilderRewardReceverPendingApproval(address builder_) external view returns (bool);
```
Comment (antomor) [inline src/interfaces/IBackersManagerRootstockCollective.sol:26]:
```suggestion
    function hasBuilderRewardReceiverPendingApproval(address builder_) external view returns (bool);
```

File: test/SetBuilderRewardReceiver.t.sol [hunk 10]
```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.20;
+
+import { BaseTest } from "./BaseTest.sol";
+import { BuilderRegistryRootstockCollective } from "../src/BuilderRegistryRootstockCollective.sol";
+import { IGovernanceManagerRootstockCollective } from "src/interfaces/IGovernanceManagerRootstockCollective.sol";
+
+contract setBuilderRewardReceiverTest is BaseTest {
```
Comment (antomor) [inline test/SetBuilderRewardReceiver.t.sol:8]:
```suggestion
contract SetBuilderRewardReceiverTest is BaseTest {
```

File: src/BuilderRegistryRootstockCollective.sol (hunk 4 above)
Comment (jmendiola222) [reply src/BuilderRegistryRootstockCollective.sol:563]: It's more of a code simplicity than a explicit requirement. It's better to join restricted rules under the same common state IMO. Anyway, a don't see a reason to allow a paused Builder to do any of this kind of changes.

File: src/BuilderRegistryRootstockCollective.sol.orig
Comment (antomor) [inline src/BuilderRegistryRootstockCollective.sol.orig:1]: I think we can delete this file

File: test/SetBuilderRewardReceiver.t.sol [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        //   THEN his original rewardReceiver address is not afected
+        assertEq(backersManager.builderRewardReceiver(builder), builder);
+        //   THEN the _newRewardReceiver address is stored on the replacement storage for him
+        assertEq(backersManager.builderRewardReceiverReplacement(builder), _newRewardReceiver);
+        //   THEN hasBuilderRewardReceiverPendingApproval returns true
+        assertEq(backersManager.hasBuilderRewardReceiverPendingApproval(builder), true);
+    }
+
+    /**
+     * SCENARIO: Builder cancels a opened rewardReceiver replacement request
```
Comment (nmagariang) [inline test/SetBuilderRewardReceiver.t.sol:104]:
```suggestion
     * SCENARIO: Builder cancels an open rewardReceiver replacement request
```

File: src/BuilderRegistryRootstockCollective.sol [hunk 12]
```diff
@@ -150,6 +156,59 @@ abstract contract BuilderRegistryRootstockCollective is CycleTimeKeeperRootstock
...
+        if (!isBuilderOperational(builder_)) revert NotOperational();
+
+        address _rewardReceiverReplacement = builderRewardReceiverReplacement[builder_];
+        if (_rewardReceiverReplacement != rewardReceiverReplacement_) revert InvalidBuilderRewardReceiver();
+        builderRewardReceiver[builder_] = _rewardReceiverReplacement;
+        emit BuilderRewardReceiverReplacementApproved(builder_, rewardReceiverReplacement_);
+    }
+
+    /**
+     * @notice returns true, if the builder has an opened request to replace his receiver address
```
Comment (nmagariang) [inline src/BuilderRegistryRootstockCollective.sol:204]:
```suggestion
     * @notice returns true if the builder has an open request to replace his receiver address
```

File: src/gauge/GaugeRootstockCollective.sol [hunk 13]
```diff
@@ -273,6 +273,11 @@ contract GaugeRootstockCollective is ReentrancyGuardUpgradeable {
         address _rewardReceiver = backersManager.builderRewardReceiver(_builder);
         if (backersManager.isBuilderPaused(_builder)) revert BuilderRewardsLocked();
         if (msg.sender != _builder && msg.sender != _rewardReceiver) revert NotAuthorized();
+        // if Builder uses the rewardReceiver account to claim, there shouldn't be an
+        // open request to replace such address, he needs to use the Builder account instead
```
Comment (nmagariang) [inline src/gauge/GaugeRootstockCollective.sol:277]:
```suggestion
        // open request to replace such address, they need to use the Builder account instead
```

File: src/interfaces/IBackersManagerRootstockCollective.sol [hunk 14]
```diff
@@ -20,6 +20,11 @@ interface IBackersManagerRootstockCollective {
      */
     function builderRewardReceiver(address builder_) external view returns (address rewardReceiver_);
 
+    /**
+     * @notice returns true, if the builder has an opened request to replace his receiver address
```
Comment (nmagariang) [inline src/interfaces/IBackersManagerRootstockCollective.sol:24]:
```suggestion
     * @notice returns true if the builder has an open request to replace his receiver address
```

File: test/GaugeRootstockCollective.t.sol [hunk 15]
```diff
@@ -971,6 +971,62 @@ contract GaugeRootstockCollectiveTest is BaseTest {
         assertEq(builder2Receiver.balance, 140 ether);
     }
 
+    /**
+     * SCENARIO: reward receiver tries to claims his rewards with a pending reward receiver address request.
```
Comment (nmagariang) [inline test/GaugeRootstockCollective.t.sol:975]:
```suggestion
     * SCENARIO: reward receiver tries to claim his rewards with a pending reward receiver address request.
```

File: test/SetBuilderRewardReceiver.t.sol [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        emit BuilderRewardReceiverReplacementCancelled(builder, builder);
+        backersManager.cancelRewardReceiverReplacementRequest();
+        //   THEN the new rewardReceiver address is back to the original
+        assertEq(backersManager.builderRewardReceiverReplacement(builder), builder);
+        //   THEN hasBuilderRewardReceiverPendingApproval returns false
+        assertEq(backersManager.hasBuilderRewardReceiverPendingApproval(builder), false);
+    }
+
+    /**
+     * SCENARIO: KYCApprover approves a opened rewardReceiver replacement request
```
Comment (nmagariang) [inline test/SetBuilderRewardReceiver.t.sol:124]:
```suggestion
     * SCENARIO: KYCApprover approves an open rewardReceiver replacement request
```

**Review by nmagariang — APPROVED** [review]
LGTM!

**Pull request merged**